src/compiler/lowering/MatMulOpLowering.cpp

```diff
@@ -14,198 +14,564 @@
+  auto a = rewriter.create<vector::SplatOp>(loc, a_single, vec_Type);
+  auto b = rewriter.create<AffineVectorLoadOp>(
+      loc, vec_Type, rhs,
+      ValueRange{fmaLoop.getInductionVar(), colLoop.getInductionVar()});
+  auto c = rewriter.create<AffineVectorLoadOp>(
+      loc, vec_Type, output,
+      ValueRange{rowLoop.getInductionVar(), colLoop.getInductionVar()});
+
+  Value res = rewriter.create<vector::FMAOp>(loc, a, b, c);
+
+  rewriter.create<AffineVectorStoreOp>(
+      loc, res, output,
+      ValueRange{rowLoop.getInductionVar(), colLoop.getInductionVar()});
+
+  // AffineYieldOp at end of loop blocks
+  rewriter.setInsertionPointAfter(fmaLoop);
+  rewriter.setInsertionPointAfter(colLoop);
+  rewriter.setInsertionPointAfter(rowLoop);
+
+  loops.push_back(rowLoop);
+  loops.push_back(colLoop);
+  loops.push_back(fmaLoop);
+  return loops;
+}
 
-    // load
-    mlir::Value a = rewriter.create<memref::LoadOp>(
-        loc, lhs, ValueRange{loopIvs[0], loopIvs[2]});
-    mlir::Value b = rewriter.create<memref::LoadOp>(
-        loc, rhs, ValueRange{loopIvs[2], loopIvs[1]});
-    mlir::Value c = rewriter.create<memref::LoadOp>(
-        loc, output, ValueRange{loopIvs[0], loopIvs[1]});
+class MatMulLowering : public OpConversionPattern<daphne::MatMulOp> {
+  const LowerMatMulOpOptions options;
+
+public:
+  using OpConversionPattern::OpConversionPattern;
+  explicit MatMulLowering(MLIRContext *context,
+                          LowerMatMulOpOptions const &options)
+      : OpConversionPattern(context, PatternBenefit(1)), options(options) {}
+
+  bool is_vectorizable(ArrayRef<int64_t> const rhsShape,
+                       Type const matrixElementType) const {
+    if (rhsShape[COL] %
+            options.getVecSize(matrixElementType.getIntOrFloatBitWidth()) !=
+        0) {
+      return false;
+    }
+    if (!matrixElementType.isa<FloatType>()) {
+      return false;
+    }
+    return true;
+  }
+
+  // TODO: Test why not
+  // Cannot tile non square matmul.
+  bool is_tileable(ArrayRef<int64_t> const rhsShape) const {
+    if (rhsShape[COL] != rhsShape[ROW]) {
+      return false;
+    }
+    return true;
+  }
+
+  LogicalResult
+  matchAndRewrite(daphne::MatMulOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto loc = op->getLoc();
+    mlir::daphne::MatrixType lhsMatrixType =
+        adaptor.getLhs().getType().dyn_cast<mlir::daphne::MatrixType>();
+    mlir::daphne::MatrixType rhsMatrixType =
+        adaptor.getRhs().getType().dyn_cast<mlir::daphne::MatrixType>();
+
+    auto lhsRows = lhsMatrixType.getNumRows();
+    auto lhsCols = lhsMatrixType.getNumCols();
+
+    auto rhsRows = rhsMatrixType.getNumRows();
+    auto rhsCols = rhsMatrixType.getNumCols();
+
+    auto matrixElementType = lhsMatrixType.getElementType();
+    // Cannot lower integer matmul, because we cannot create Memrefs with
+    // integer type at the moment.
```

Comment:
Integer memrefs should generally work, see `src/compiler/lowering/EwOpsLowering.cpp`. Handle this as already discussed in our meeting, so similarly to the `TODO` above.

Reply:
The original problem was three fold. 1. Filling the output memref with correctly signed 0 integer values, when we so far only used arith::ConstantOps which are MLIR unsigned. 2. Similarly arith::AddIOp only operates on signless types. 3. SumAllOp could not handle integer types for the same reasons.
In EwOps this is handled by casting from signfull to signless where necessary. I took this approach and applied it to MatMulLowering and SumAllOpLowering.